### ISSUE TITLE
Email Marketing integration issues with Constant Contact

### ISSUE THREAD

**Issue by jrrdnx** [solspace/craft3-freeform#616]
**Describe the bug or issue you're experiencing**

I'm able to complete the process of setting up a Constant Contact email marketing integration as outlined here: https://docs.solspace.com/craft/freeform/v3/integrations/email-marketing/constant-contact.html

All setup steps have been completed successfully and the edit screen shows the green circle with "Authorized" above the Redirect URI.

However, when attempting to set up the integration on an existing or new form, clicking the "Refresh lists" button changes to "Refreshing" for a period but the Mailing Lists dropdown is never updated.

Checking the Freeform error logs shows several "Too many requests" entries:

```
[2023-03-02T11:47:03.968181-05:00] Email Marketing Integration.ERROR: {"error_key":"throttled","error_message":"Too Many Requests"} {"exception":"Client error: `GET https://api.cc.email/v3/contact_custom_fields` resulted in a `429 Too Many Requests` response:\n{\"error_key\":\"throttled\",\"error_message\":\"Too Many Requests\"}\n"} []
```

I wasn't sure if there's been a change with the Constant Contact API recently, an account plan/billing issue with the Constant Contact account (currently on the 'Email Plus' plan), an issue on Freeform's end, or on my end with the initial setup, but figured I'd start here.

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 3.7.67
- Freeform Version: 3.13.25
- Freeform Edition: Pro
- Fresh Install or Upgrade: Initially installed Lite, recently upgrade to Pro

**Comment by jrrdnx** [comment]
Good morning, wanted to check and see if you had any thoughts on this issue? Thanks!

**Comment by kjmartens** [comment]
Hi @jrrdnx,

I'm very sorry for the delay.

I haven't seen this issue surface for other customers yet. Can you confirm what your setup is like? E.g. 

1. How many email lists do you have?
2. How many Constant Contact fields do you have?
3. Anything else that might be helpful. 🙂

**Comment by jrrdnx** [comment]
Thanks @kjmartens !

It's a client's CC account so our access is limited, but if you need more details I can coordinate that (they're using 2FA, so takes some planning in timing). I do know that they are currently on the 'Email Plus' plan.

1. "We have well over 50 lists, if not close to 100. But only a small handful of them are tied to the subscription forms."
2. I'm not sure of the overall number of fields, but the list we need to connect the form to is only using 4: first name, last name, email address, and birthday.

**Comment by kjmartens** [comment]
Hi @jrrdnx,

I did some more testing on my end and can't get the issue to happen for me. However, I wonder if it might have something to do with there being so many lists. The error does mention `contact_custom_fields`, so I wonder if their CC account actually contains a large number of custom fields. I understand that they only wish to map a few, but the Freeform integration will still have to attempt to grab all available fields across the lists in order to configure field mapping. Can you check how many fields they have in CC? 🙂

**Comment by jrrdnx** [comment]
@kjmartens 

I have verified they are currently sitting at 84 lists, and if I'm looking in the correct location (under the Contacts menu > button with the 3 dots > Manage Custom Fields) they only have 3 custom fields.

**Comment by jrrdnx** [comment]
Wanted to note that I updated to Craft Pro 3.8.5 and Freeform Pro 3.13.26 but the issue persists.

Opening an existing form with an email marketing field that triggers the API request results in two of the '429 Too Many Requests' errors logged within 6-12 seconds of each other.

**Comment by jrrdnx** [comment]
@kjmartens 

I had a little free time to dig into the plugin code and try to debug what could be causing this. It appears that it's making a new API call to get the custom fields for each list, which in this case for 84 lists is clearly hitting the max of 4 requests per second.

The `fetchLists()` function in `solspace/craft-freeform/packages/plugin/src/Integrations/MailingLists/ConstantContact3.php:247` is calling `$this->fetchFields()` for each list, but this doesn't appear to be necessary. There is a `$listId` argument for that `fetchFields()` function but it doesn't appear to actually be used anywhere.

In our dev environment, I modified the `fetchLists()` function to use a single call to `fetchFields()` before it loops through all of the lists (`$fields = $this->fetchFields(0);`, passing 0 since there is no List ID to pass), and using this variable when instantiating the new ListObject on line 287:

```php
$lists[] = new ListObject(
    $this,
    $list->list_id,
    $list->name,
    $fields
);
```

With this change the dropdown of Mailing Lists is successfully populated when editing a form, and I'm able to select the Mailing List and map fields without issue.

I'd be happy to submit a pull request with this fix, but wasn't sure about the $listId argument for fetchFields(). It makes sense that it would need to fetch the custom fields used with each list, but it doesn't appear to achieve that functionality. At the least, I think this could be delayed and fired once when the selected list is chosen from the Mailing Lists dropdown to alleviate the large number of API calls up front for users with a large number of lists. Thoughts?

**Comment by kjmartens** [comment]
Thanks @jrrdnx. We'll check into this shortly. 🙂

**Comment by kjmartens** [comment]
This should now be resolved in Freeform 3.13.27 and 4.0.20 🙂

**Comment by jrrdnx** [comment]
Thank you @kjmartens !
